File: ios/MiniBozor/Data/Repositories/OrderRepository.swift
import Foundation

/// Screens 19–29, 32–33.
struct OrderRepository {
    private let api = APIClient.shared

    // Addresses

    func addresses() async -> Outcome<[AddressDTO]> {
        await run { try await api.get("addresses") }
    }

    func createAddress(_ body: AddressRequest) async -> Outcome<AddressDTO> {
        await run { try await api.post("addresses", body: body) }
    }

    func updateAddress(id: Int, body: AddressRequest) async -> Outcome<AddressDTO> {
        await run { try await api.put("addresses/\(id)", body: body) }
    }

    func deleteAddress(id: Int) async -> Outcome<Void> {
        await run { let _: MessageDTO = try await api.delete("addresses/\(id)") }
    }

    // Delivery

    func slots(days: Int = 3) async -> Outcome<[SlotDayDTO]> {
        await run {
            try await api.get("delivery/slots", query: [URLQueryItem(name: "days", value: String(days))])
        }
    }

    func pickupPoints() async -> Outcome<[PickupPointDTO]> {
        await run { try await api.get("delivery/pickup-points") }
    }

    // Cards

    func cards() async -> Outcome<[CardDTO]> {
        await run { try await api.get("payment-cards") }
    }

    func addCard(_ body: CardRequest) async -> Outcome<CardDTO> {
        await run { try await api.post("payment-cards", body: body) }
    }

    func makeCardDefault(id: Int) async -> Outcome<CardDTO> {
        await run { try await api.post("payment-cards/\(id)/default") }
    }

    func deleteCard(id: Int) async -> Outcome<Void> {
        await run { let _: MessageDTO = try await api.delete("payment-cards/\(id)") }
    }

    // Checkout and orders

    func preview(_ body: CheckoutRequest) async -> Outcome<CheckoutPreviewDTO> {
        await run { try await api.post("checkout/preview", body: body) }
    }

    func place(_ body: CheckoutRequest) async -> Outcome<OrderDTO> {
        await run { try await api.post("orders", body: body) }
    }

    func orders(active: Bool?, page: Int = 1) async -> Outcome<PageDTO<OrderSummaryDTO>> {
        var items = [URLQueryItem(name: "page", value: String(page))]
        if let active { items.append(URLQueryItem(name: "active", value: active ? "true" : "false")) }
        return await run { try await api.get("orders", query: items) }
    }

    func order(id: Int) async -> Outcome<OrderDTO> {
        await run { try await api.get("orders/\(id)") }
    }

    func cancel(id: Int, body: CancelRequest) async -> Outcome<OrderDTO> {
        await run { try await api.post("orders/\(id)/cancel", body: body) }
    }

    func requestReturn(id: Int, body: ReturnRequestBody) async -> Outcome<ReturnDTO> {
        await run { try await api.post("orders/\(id)/return", body: body) }
    }

    func returns() async -> Outcome<[ReturnDTO]> {
        await run { try await api.get("returns") }
    }

    func cancelReasons() async -> Outcome<[ReasonDTO]> {
        await run { try await api.get("orders/reasons/cancel") }
    }

    func returnReasons() async -> Outcome<[ReasonDTO]> {
        await run { try await api.get("orders/reasons/return") }
    }
}

File: android/app/src/main/java/uz/minibozor/data/repository/OrderRepository.kt
package uz.minibozor.data.repository

import uz.minibozor.core.util.Outcome
import uz.minibozor.core.util.apiCall
import uz.minibozor.data.remote.MiniBozorApi
import uz.minibozor.data.remote.dto.*
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class OrderRepository @Inject constructor(private val api: MiniBozorApi) {

    suspend fun addresses(): Outcome<List<AddressDto>> = apiCall { api.addresses() }

    suspend fun createAddress(body: AddressRequest): Outcome<AddressDto> =
        apiCall { api.createAddress(body) }

    suspend fun updateAddress(id: Int, body: AddressRequest): Outcome<AddressDto> =
        apiCall { api.updateAddress(id, body) }

    suspend fun deleteAddress(id: Int): Outcome<Unit> =
        apiCall { api.deleteAddress(id) }.let { if (it is Outcome.Failure) it else Outcome.Success(Unit) }

    suspend fun slots(days: Int = 3): Outcome<List<SlotDayDto>> = apiCall { api.slots(days) }

    suspend fun pickupPoints(): Outcome<List<PickupPointDto>> = apiCall { api.pickupPoints() }

    suspend fun cards(): Outcome<List<CardDto>> = apiCall { api.cards() }

    suspend fun addCard(body: CardRequest): Outcome<CardDto> = apiCall { api.addCard(body) }

    suspend fun makeCardDefault(id: Int): Outcome<CardDto> = apiCall { api.makeCardDefault(id) }

    suspend fun deleteCard(id: Int): Outcome<Unit> =
        apiCall { api.deleteCard(id) }.let { if (it is Outcome.Failure) it else Outcome.Success(Unit) }

    suspend fun preview(body: CheckoutRequest): Outcome<CheckoutPreviewDto> =
        apiCall { api.checkoutPreview(body) }

    suspend fun place(body: CheckoutRequest): Outcome<OrderDto> = apiCall { api.createOrder(body) }

    suspend fun orders(active: Boolean?, page: Int = 1): Outcome<PageDto<OrderSummaryDto>> =
        apiCall { api.orders(active, page) }

    suspend fun order(id: Int): Outcome<OrderDto> = apiCall { api.order(id) }

    suspend fun cancel(id: Int, body: CancelRequest): Outcome<OrderDto> =
        apiCall { api.cancelOrder(id, body) }

    suspend fun requestReturn(id: Int, body: ReturnRequestBody): Outcome<ReturnDto> =
        apiCall { api.requestReturn(id, body) }

    suspend fun returns(): Outcome<List<ReturnDto>> = apiCall { api.returns() }

    suspend fun cancelReasons(): Outcome<List<ReasonDto>> = apiCall { api.cancelReasons() }

    suspend fun returnReasons(): Outcome<List<ReasonDto>> = apiCall { api.returnReasons() }
}
